plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -229,81 +228,80 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
         List<ConnectorSplit> splits = new ArrayList<>(maxSize);
         while (splits.size() < maxSize && (fileTasksIterator.hasNext() || fileScanIterator.hasNext())) {
             if (!fileTasksIterator.hasNext()) {
-                FileScanTask wholeFileTask = fileScanIterator.next();
-                if (wholeFileTask.deletes().isEmpty() && noDataColumnsProjected(wholeFileTask)) {
-                    fileTasksIterator = List.of(wholeFileTask).iterator();
+                if (limit.isPresent() && limit.getAsLong() <= outputRowsLowerBound) {
+                    finish();
+                    continue;
                 }
-                else {
-                    fileTasksIterator = wholeFileTask.split(targetSplitSize).iterator();
+                FileScanTask wholeFileTask = fileScanIterator.next();
+                boolean fileHasAnyDeletions = !wholeFileTask.deletes().isEmpty();
```

Comment:
nit: we always use this file in this form `!fileHasAnyDeletions` . Maybe rename it to `fileHasNoDeletions`